PackedByteArray: Direct big-endian encoding/decoding

cruglet opened the pull request:
Added an alternate method for all encoding/decoding functions above u8 in order to read/write in big endian. Also slightly cleaned up a bit of the encoding/decoding logic.
- This aims to help read & write BE binary without having to manually implement those functions in GDScript. Existing implementations are unaffected as this is another method entirely.

___

- *Production edit: Related to https://github.com/godotengine/godot/pull/105178
and https://github.com/godotengine/godot-docs/issues/10848.*

bruvzg commented on core/variant/variant_call.cpp:
```diff
@@ -2530,15 +2530,15 @@ static void _register_variant_builtin_methods_array() {
-	bind_function(PackedByteArray, decode_half, _VariantCall::func_PackedByteArray_decode_half, sarray("byte_offset"), varray());
-	bind_function(PackedByteArray, decode_float, _VariantCall::func_PackedByteArray_decode_float, sarray("byte_offset"), varray());
-	bind_function(PackedByteArray, decode_double, _VariantCall::func_PackedByteArray_decode_double, sarray("byte_offset"), varray());
+	bind_function(PackedByteArray, decode_u16, _VariantCall::func_PackedByteArray_decode_u16, sarray("byte_offset", "big_endian"), varray(false));
+	bind_function(PackedByteArray, decode_s16, _VariantCall::func_PackedByteArray_decode_s16, sarray("byte_offset", "big_endian"), varray(false));
+	bind_function(PackedByteArray, decode_u32, _VariantCall::func_PackedByteArray_decode_u32, sarray("byte_offset", "big_endian"), varray(false));
+	bind_function(PackedByteArray, decode_s32, _VariantCall::func_PackedByteArray_decode_s32, sarray("byte_offset", "big_endian"), varray(false));
+	bind_function(PackedByteArray, decode_u64, _VariantCall::func_PackedByteArray_decode_u64, sarray("byte_offset", "big_endian"), varray(false));
+	bind_function(PackedByteArray, decode_s64, _VariantCall::func_PackedByteArray_decode_s64, sarray("byte_offset", "big_endian"), varray(false));
+	bind_function(PackedByteArray, decode_half, _VariantCall::func_PackedByteArray_decode_half, sarray("byte_offset", "big_endian"), varray(false));
+	bind_function(PackedByteArray, decode_float, _VariantCall::func_PackedByteArray_decode_float, sarray("byte_offset", "big_endian"), varray(false));
+	bind_function(PackedByteArray, decode_double, _VariantCall::func_PackedByteArray_decode_double, sarray("byte_offset", "big_endian"), varray(false));
```
This requires compatibility methods for GDExtension (and it is not supported for built-in methods), so the only option right now is adding new methods instead (`decode_u16be` or something like this).

cruglet replied:
> This requires compatibility methods for GDExtension (and it is not supported for built-in methods), so the only option right now is adding new methods instead (`decode_u16be` or something like this).

Good catch, BE logic has been made separate now

cruglet committed:
PackedByteArray: Big-endian encoding/decoding

Added respective functions for directly reading/writing from big-endian ordered bytes.
- This aims to help read & write BE binary without having to manually implement those functions in GDScript. Existing implementations are unaffected as these are supplementary functions.

cruglet commented:
Alright, I think this should cover it. Let me know if I missed anything 👍

Calinou reviewed (commented):
Tested locally with https://github.com/Calinou/slippi-hitsounds (I replaced uses of `bswap` with the new methods), it works as expected.

I suggest adding unit tests to cover the new methods in `tests/test_marshalls.h` (you can add them at the end of each test case).

cruglet commented:
> I suggest adding unit tests to cover the new methods in tests/test_marshalls.h (you can add them at the end of each test case).

Unit tests have been appended to their LE counterparts 👍

Calinou reviewed (approved):
Tests look great :slightly_smiling_face:

Splizard commented:
Is this still necessary? now that bswap16, bswap32 and bswap64 are included in Godot 4.5 (these are intended for working with big endian data)
https://docs.godotengine.org/en/stable/classes/class_packedbytearray.html#class-packedbytearray-method-bswap16

AThousandShips commented on doc/classes/PackedByteArray.xml:
```diff
@@ -259,6 +346,13 @@
+			<return type="void" />
+			<param index="0" name="byte_offset" type="int" />
+			<param index="1" name="value" type="int" />
+			<description>
```
Needs a description

cruglet commented:
> Is there any practical use for this?

See above
> I still think so. bswap gives low level swapping but doesn’t on its own necessarily give direct encoding/decoding or automatic integration in PackedByteArray. These `encode_x_be` methods just serve as complimentary BE implementations for the current `encode_x` methods available, and make the intention much clearer and less verbose

The main issue this addresses is that the alternative to decode the first four bytes of data in BE from a PackedByteArray would be something along the lines of:
```
array.bswap32().decode_u32()
```
And it gets worse if you wanted it to be non-destructive:
```
array.slice(0, 4).bswap32().decode_u32()
```
So the new methods help make this much more readable and intuitive: 
```
array.decode_u32_be()
```
Throughout the development of 4.5 and forward I've seen pushes towards more explicit distinctions between big endian and little endian operations (FileAccess in particular), so this would just be an extension of that.

Splizard commented:
@cruglet
Have you seen the PR for bswap? 
https://github.com/godotengine/godot/pull/105145

You've mentioned bswap provides low level swapping, I'm not entirely sure what you mean by this, as the PR shows that these methods were specifically introduced for working with big endian data.

Your changes certainly provide an alternative way to work with big endian data, you present a strong case for how it's more concise than bswap. This comes at the cost of introducing many more methods, and making bswap mostly redundant.

If you still believe it's best to merge this PR, I think it needs a compelling case that demonstrates a wide-enough need that outweighs the inconvenience of using bswap.

bruvzg commented:
> See above

I do not see any mentions of real world use cases.

This might be a bit cleaner than using `bswap` (which itself only exist to swap large chunks of data without performance impact, for something `slice(0, 4).bswap32()` it can be done by manually swapping bytes), but if it's used only for a few obscure file formats, it probably is not worth adding core functions.

cruglet commented:
> You've mentioned bswap provides low level swapping, I'm not entirely sure what you mean by this, as the PR shows that these methods were specifically introduced for working with big endian data. 

The PR "partially adressess" the problem. `bswap` is a low-level utility for swapping byte order, but it doesn’t change the fact that the encode_x and decode_x methods operate in little endian by default. The main issue with `bswap` in this case is that it blindly swaps bytes in fixed-width chunks without any awareness of the semantic structure of the data.

> Your changes certainly provide an alternative way to work with big endian data, you present a strong case for how it's more concise than bswap. This comes at the cost of introducing many more methods, and making bswap mostly redundant.
>
> If you still believe it's best to merge this PR, I think it needs a compelling case that demonstrates a wide-enough need that outweighs the inconvenience of using bswap.

I would also like to echo some points brought up in a related issue linked to the bswap PR (cc @YuriSizov):
- Mutating data just to work around the lack of endianness control in encode_x/decode_x isn’t ideal.
- Currently, bswap doesn’t really offer more than what could be done with a reverse and a copy of the section you want to read/write, except maybe for very large homogeneous data chunks. This very quickly breaks down when working with buffers that have multiple different data types in one stream, let alone if some of that mixed data isn't necessarily BE.
- One offered compromise was to add the flag to all encode_* and decode_* methods directly (which was originally done in my PR), or to introduce new methods for BE. I shifted to the latter in this case.


> > See above
> 
> I do not see any mentions of real world use cases.
> 
> This might be a bit cleaner than using `bswap` (which itself only exist to swap large chunks of data without performance impact, for something `slice(0, 4).bswap32()` it can be done by manually swapping bytes), but if it's used only for a few obscure file formats, it probably is not worth adding core functions.

I get where this is coming from, but I feel as though this conflates frequency of use in specific cases with API completeness and correctness. Claiming this is bloat overlooks that `bswap` and BE encode/decode serve fundamentally different purposes. In fact, the very existence of a dedicated `bswap` method acknowledges that working with different endianness types is important enough to warrant core support. This BE encode/decode extension is simply a complementary & safe way to fill that gap, not necessarily redundant or niche.

Splizard commented:
If you are worried about lines of code and overhead, perhaps we should consider adding bswap functions to @GlobalScope?

```gdscript
Variant bswap16(x: Variant)
Variant bswap32(x: Variant)
Variant bswap64(x: Variant)

# returns x with the underlying byte order swapped, for the least N significant bits, useful
# for converting to big endian values from little endian values and vice versa. Floats
# are converted to and from their N-bit representation, before being swapped.
# Supported types: int, float

bswap32(decode_u32(offset)) # decodes a big endian uint32 at offset
```

As @bruvzg has mentioned, I don't think anybody is trying to argue here that bswap methods are more intuitive, but it's also a reality that big-endian data is primarily a special case concern today. This is why there is some hesitation about adding 18 new methods and an explicit decision point (particularly, by newcomers who are just trying to serialize some game data) about which one to use.

Everyone seems to agree that adding a toggle (that defaults to little-endian) to the PackedByteArray (or its methods) is another good solution and this may be possible in the future if builtin methods get compatibility hashes for GDExtension, or the underlying data structure is changed. Merging this now would realistically shut down a good alternative option for dealing with big-endian data until something like Godot 5.

Splizard commented:
That seems like an option, FileAccess already looks a lot like a BinaryBuffer (except tied to a particular file) and one of the annoying things about PackedByteArrays is tracking offsets.

Would it be possible to introduce something like a virtual BinaryBuffer/BinaryAccess class that FileAccess simply inherits its existing data-manipulation methods from? Would this be backwards compatible (enough)? Then you could have something like FileAccess and MemoryAccess and GDExtensions could provide their own BinaryBuffer/BinaryAccess implementations too, ie. exposing certain memory structures directly.
